src/broadcast/broadcast.rs

```diff
@@ -274,14 +275,14 @@ impl<N: NodeIdT> Broadcast<N> {
         let hash = *p.root_hash();
 
         // Save the proof for reconstructing the tree later.
-        self.echos.insert(sender_id.clone(), p);
+        self.echos.insert(sender_id.clone(), (hash.clone(), Some(p.clone())));
 
         // Upon receiving `N - 2f` `Echo`s with this root hash, send `CanDecode`
-        if !self.can_decode_sent && self.count_echos(&hash) >= self.netinfo.num_correct() - self.netinfo.num_faulty() {
+        if !self.can_decode_sent && self.count_echos_full(&hash) >= self.netinfo.num_correct() - self.netinfo.num_faulty() {
```

Comment:
Shouldn't this be `self.netinfo.num_correct() - 2 * self.netinfo.num_faulty()`? Or, even better, `self.coding.data_shard_num()`? (That makes the meaning more clear: We have enough shards to decode the data.)

Reply:
You are right. `self.coding.data_shard_num()` is better.
Will fix this.